File: 2025/07/main.go
package main

import (
	"fmt"
	"strings"

	"2025/lib"
)

const SPLITTER = '^'

type Stream struct {
	y       int
	x       int
	counter int
}

func (s *Stream) down() {
	s.y++
}

func (s *Stream) left() {
	s.x--
}

func (s *Stream) right() {
	s.x++
}

func (s *Stream) updateCounter(num int) {
	s.counter = num
}

func removeDups(ss []*Stream) []*Stream {
	for i := 0; i < len(ss)-1; i++ {
		for j := i + 1; j < len(ss); j++ {
			a := ss[i]
			b := ss[j]
			if a.y == b.y && a.x == b.x {
				a.updateCounter(a.counter + b.counter)
				ss = append(ss[:j], ss[j+1:]...)
				j--
			}
		}
	}
	return ss
}

func solve() {
	// text := lib.ReadInput("07/example.txt")
	text := lib.ReadInput("07/input.txt")
	rows := lib.GetRows(text)
	splitsAmount := 0
	startIndex := strings.Index(rows[0], "S")
	ss := []*Stream{ // need * because we want to array of pointers
		{0, startIndex, 1},
	}

	for ss[0].y != len(rows)-1 {
		newSs := []*Stream{}
		for _, s := range ss {
			s.down()
			if rows[s.y][s.x] == SPLITTER {
				splitsAmount++
				s2 := &Stream{s.y, s.x, s.counter}
				s.left()
				s2.right()
				newSs = append(newSs, s2)
			}
		}
		ss = append(ss, newSs...)
		ss = removeDups(ss)
	}
	sum := 0
	for _, s := range ss {
		sum += s.counter
	}
	fmt.Println("splits:", splitsAmount)
	fmt.Println("sum:", sum)
}

func main() {
	solve()
}

File: 2025/07/main.ts
import fs from 'node:fs/promises';
// const fs = require('node:fs/promises');

async function read(path: string) {
  const data = await fs.readFile(path, { encoding: 'utf8' });
  return data
}

class Beam {
  y: number = -1
  x: number = -1
  counter: number = 1

  constructor(y: number, x: number, counter: number) {
    this.y = y
    this.x = x
    this.counter = counter
  }

  updateCounter(num: number) {
    this.counter = num
  }
  moveDown() {
    this.y++
  }
  moveLeft() {
    this.x--
  }
  moveRight() {
    this.x++
  }
}

function removeDups(beams: Beam[]) {
  for (let i = 0; i < beams.length - 1; i++) {
    for (let j = i + 1; j < beams.length; j++) {
      if (beams[i].y === beams[j].y && beams[i].x === beams[j].x) {
        beams[i].updateCounter(beams[i].counter + beams[j].counter)
        beams = beams.filter((_, index) => index !== j)
        j--
      }
    }
  }
  return beams
}

function solveAllParts(data: string[][]) {
  let startIndex = data[0].indexOf('S')
  let beams = [new Beam(0, startIndex, 1)]
  let split: { y: number, x: number }[] = []

  while (beams[0].y !== data.length - 1) {
    // move down
    let newBeams: Beam[] = []
    for (let b of beams) {
      b.moveDown()
      if (data[b.y][b.x] === '^') {
        split.push({ y: b.y, x: b.x, })
        let b2 = new Beam(b.y, b.x, b.counter)
        b.moveLeft()
        b2.moveRight()
        newBeams.push(b2)
      }
    }
    // combine
    beams = [...beams, ...newBeams]
    // remove dups
    beams = removeDups(beams)
  }

  let sum = 0
  for (let b of beams) {
    sum += b.counter
  }

  console.log('splits amount:', split.length)
  console.log('sum:', sum)
}

async function main() {
  // let data: string = await read('example.txt')
  let data: string = await read('input.txt')
  let arr = data.split('\n').map(row => row.split(''))
  solveAllParts(arr)
}

main()
